fix: refresh kubernetes metadata watcher on stream error

jakedipity opened the pull request:
In some cases, for unknown reasons, the k8s metadata watcher stream can
become inactive after receiving an error. This causes the cache to
become stale over time and for labels and annotations to be stale or
completely missing.

This PR fixes the issue by taking direct control of error handling on
stream errors and recreates the metadata watcher to start fresh.

Ref: LOG-16194

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -162,22 +163,48 @@ pub async fn _main(
-                            error!("{}", message);
-                            panic!("{}", message);
+                    let k8s_client_clone = k8s_client.clone();
+                    tokio::spawn(async move {
```
client is reused? connection is not recreated?
and we reuse the storage?

jakedipity replied:
Ah I didn't think about the client, it makes sense to recreate it every time in case our issue is arising there. The reflector store is dropped with the thread each refresh; so a new store and watcher is created each time.

dkhokhlov replied:
I assume while we restart watcher we will not lose notifications. could you check API spec that we do get some backlog of notifications after new connect?

jakedipity commented on bin/src/_main.rs:
```diff
@@ -162,22 +163,48 @@ pub async fn _main(
+                                }
+                            };
+
+                            tokio::time::sleep(tokio::time::Duration::from_secs(10 * 60)).await;
```
Yeah I can move this to a constant. I picked 10 minutes arbitrarily, let me know if you think we should change it.

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -51,6 +52,7 @@ use crate::stream_adapter::{StrictOrLazyLineBuilder, StrictOrLazyLines};
 pub static REPORTER_LEASE_NAME: &str = "logdna-agent-reporter-lease";
 pub static K8S_EVENTS_LEASE_NAME: &str = "logdna-agent-k8-events-lease";
 pub static DEFAULT_CHECK_FOR_LEADER_S: i32 = 300;
+pub static DEFAULT_K8S_WATCHER_REFRESH: Duration = Duration::from_secs(10 * 60);
```
DEFAULT_K8S_WATCHER_REFRESH_S
?

jakedipity replied:
Renamed to hopefully be clearer.

jakedipity commented:
I moved the `Mutex` to only surround the reflector inside of `K8sMetadata` instead of around the entire middleware itself. This simplifies things some and helps limit the critical section.

The `kick_over` function now returns a new `StoreSwapIntent` instead of the reflector store itself; this is to make it clear what the object is supposed to be used for.

I added a `VecDeque` to track more than one running  watcher thread. This collection will only ever have at most 2 watcher threads and allows us to overlap the running threads, so a new watcher can populate it's store before being utilized by processing lines in the middleware.

jakedipity commented:
@dkhokhlov Should be ready for re-review

Also @c-nixon would be good to have some extra eyes on this

jakedipity commented:
@c-nixon Bumped kube to latest, `0.80.0`. Needed to update our minimum kubernetes version to `v1.20`; however, and needed to make some small tweaks because of some api changes.

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -139,110 +142,170 @@ pub async fn _main(
+                                debug!("k8s metadata watcher kicked over!");
+                                let abortable_stream = Abortable::new(stream, abort_registration);
+                                (
+                                    tokio::spawn(async move {
```
I do not feel good about this additional spawn inside loop.  how about using a channel instead of abortable stream?

jakedipity replied:
outdated

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -139,110 +142,170 @@ pub async fn _main(
-                        )
-                        .await;
+                        let (stream_handle, abort_registration) = AbortHandle::new_pair();
+                        let (thread_handle, swap_intent) = match middleware
```
first loop run starts with kick over right away?

jakedipity replied:
outdated

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -139,110 +142,170 @@ pub async fn _main(
+                        };
+
+                        handle_queue.push_back((thread_handle, stream_handle, Some(swap_intent)));
+                        if handle_queue.len() == 2 {
```
hard coded 2. is it when queue is full?

jakedipity replied:
outdated - the new code now uses a state machine of sorts

dkhokhlov commented on bin/src/_main.rs:
```diff
@@ -139,110 +142,170 @@ pub async fn _main(
+                            debug!("delaying k8s metadata watcher rotation so new store can populate...");
+                            // when rotating watchers, introduce a small delay so the new
+                            // watcher has a chance to populate its cache
+                            sleep(K8S_WATCHER_ROTATION_DELAY).await;
```
how about to use a channel here? for explicit handshake.

jakedipity replied:
outdated

dkhokhlov commented on bin/tests/it/k8s.rs:
```diff
@@ -28,7 +28,7 @@ async fn print_pod_logs(client: Client, namespace: &str, label: &str) {
                 let mut logs = pods
                     .log_stream(
-                        &p.name(),
+                        &p.metadata.name.clone().unwrap(),
```
always Some()?

jakedipity replied:
Not sure, with the update to our rust dependencies, these values are now returned as an Option. IMO given that it's a test I think it's okay to just unwrap without trying to handle the error - panicking will be seen as a test failure.

dkhokhlov commented on common/k8s/src/lib.rs:
```diff
@@ -76,7 +75,8 @@ pub fn create_k8s_client_default_from_env(
     }
 
-    let config = Config::from_cluster_env()?;
+    let config = Config::incluster_dns()?;
```
why renamed to dns?

jakedipity replied:
Because we are using rustls-tls, we have to use the dns instead of from the env: https://docs.rs/kube/latest/kube/struct.Config.html#method.incluster_dns

dkhokhlov commented on common/k8s/src/middleware/metadata.rs:
```diff
@@ -82,56 +105,51 @@ impl K8sMetadata {
+                    Err(e) => warn!("k8s watch stream error: {}", e),
+                })
+                .take_while(|x| std::future::ready(x.is_ok()))
+                .map(Result::unwrap),
```
safe unwrap?

jakedipity replied:
Yes, `take_while` will terminate the stream if an error is encountered, so the unwrap will never get an error.

dkhokhlov commented on common/k8s/src/middleware/metadata.rs:
```diff
@@ -148,38 +166,40 @@ impl Middleware for K8sMetadata {
-                            )
-                            .is_err()
-                        {
+                if let Some(ref store) = *self.store.lock().unwrap() {
```
always success unwrap?

jakedipity replied:
Unwrap could panic if a previous holder of the lock panicked while holding the lock. There's not much else we can do in this situation, so panicking seems like an acceptable result.

dkhokhlov commented on common/k8s/src/middleware/runner.rs:
```diff
@@ -0,0 +1,111 @@
+                error!("{}", message);
+                panic!("{}", message);
+            }
+            State::Kicking(first, middleware_k8s_client.unwrap())
```
middleware_k8s_client always Some()? Err or Option?
may be match...

jakedipity replied:
The error case is handled in the lines before (it'll panic), so the unwrap should always have a `Some`. I simplified the code to process the error at the same time it unwraps it.

jakedipity commented on common/middleware/src/lib.rs:
```diff
@@ -17,6 +17,41 @@ pub trait Middleware: Send + Sync + 'static {
     fn process<'a>(&self, lines: &'a mut dyn LineBufferMut) -> Status<&'a mut dyn LineBufferMut>;
 }
 
+pub enum MixedMiddleware<T: Middleware> {
```
I implemented this suggestion.

jakedipity commented on common/k8s/src/middleware/runner.rs:
```diff
@@ -0,0 +1,111 @@
+    Running(JoinHandle<()>),
+}
+
+async fn step_trampoline<'a>(
```
I added exponential backoff, it will pause after the Running state. It starts at 2 seconds and goes to a maximum of 30 seconds. If the running state is able to run successfully for longer than 30 seconds, the attempts counter will get reset and go back to a 2 second delay.
